### PR TITLE
 Switch to DynamicPPL

### PR DESCRIPTION
**Pull request by mohamed82008**
This is a WIP PR that makes Turing depend on DynamicPPL.

### CODE REVIEW

**Comment by mohamed82008** [comment]
There seems to be issues with the new AdvancedHMC and HMCDA in Turing. CC: @xukai92

**Comment by mohamed82008** [comment]
I am getting the same failures on master locally.

**Comment by xukai92** [comment]
Since which AHMC version we have this error?

**Comment by mohamed82008** [comment]
I will check.

**Comment by mohamed82008** [comment]
NUTS also seems to have issues.

**Comment by mohamed82008** [comment]
Since 0.2.15

**Comment by xukai92** [comment]
Thanks for checking. I will work on it when I'm with my laptop later.

**Comment by xukai92** [comment]
It's weird that I cannot reproduce the error locally.

**Comment by xukai92** [comment]
> It's weird that I cannot reproduce the error locally.

Sorry my tests were influnced by the Manifest.toml file - since when we had it back?

**Comment by mohamed82008** [comment]
> Sorry my tests were influnced by the Manifest.toml file - since when we had it back?

We don't it's gitignored, it is just generated automatically on your system when you add a dependency using `Pkg` in the package's environment.

**Comment by mohamed82008** [comment]
Or if you instantiate the package's environment.

**Comment by xukai92** [comment]
Oh you are right. Thanks for sharing this :) I can reproduce the error now and let me figure out what's going on here.

**Comment by xukai92** [comment]
I pushed some commits to fix deprecations. The error will need https://github.com/TuringLang/AdvancedHMC.jl/pull/156. Basically I forget to check cases where `n_adapts` is smaller than the default adpataion buffer of the Stan HMC adaptor and this causes the error.

**Comment by xukai92** [comment]
We might not need https://github.com/TuringLang/Turing.jl/pull/1042/commits/ffbc169371bf915b115cae02cac7fed7219297f7 to fix the error any more but I think it's good to have it. How do you think?

Also cc @cpfiffer

**Comment by cpfiffer** [comment]
ffbc169 may have the unintended consequence of not allowing us to automatically set the number of adaptation samples, which might be fine. Currently, if the number of adaptation samples is set to zero, it'll be changed to something reasonable in `sample_init!`:

https://github.com/TuringLang/Turing.jl/blob/ffbc169371bf915b115cae02cac7fed7219297f7/src/inference/hmc.jl#L125

It's probably fine to do this kind of thing, but it makes convenience functions like just `NUTS()` use no adaptation by default -- is that something we'd prefer?

File: src/utilities/stan-interface.jl [hunk 1]
```diff
@@ -66,6 +66,7 @@ function sample(
     end
 end
 
+# FIXME: where is `spl` in the code below?
```
Comment (mohamed82008) [inline src/utilities/stan-interface.jl]: Seriously where is `spl` in the code below? 😆

Comment (mohamed82008) [reply src/utilities/stan-interface.jl]: Is this function even used?

**Comment by xukai92** [comment]
I think the current order of automatically settings the number of adaptation is flawed. What would happen if one is intended to set it to `0`? The current behaviour will still change it to something else. 

I think what we had before could handle it correctly, i.e. setting it during the construction of algorithm.

File: src/utilities/stan-interface.jl (hunk 1 above)
Comment (xukai92) [reply src/utilities/stan-interface.jl]: Might not be correctly tested since whatever.

**Comment by cpfiffer** [comment]
> I think the current order of automatically setting the number of adaptation is flawed. What would happen if one is intended to set it to `0`? The current behaviour will still change it to something else.

Yeah, but this was also our old logic carried in from the original Hamiltonian implementations -- I mostly just moved it around. 

The problem with handling this during the construction of the algorithm is that currently the inference algorithm constructor does not know the total number of iterations because `N` is part of the outer `sample` call. We could probably get around that, though.

Maybe a better way to do this would be to move the number of adapatation samples to the keyword side:

```julia
sample(model, NUTS(500, 0.65), 1000)
```

would become

```julia
sample(model, NUTS(0.65), 1000; n_adapts=500)
```

Then we could handle all this more robustly at the inference constructor by passing keywords around. Having the number of adaptations in the inference constructor always seemed a little weird to me anyway.

**Comment by mohamed82008** [comment]
I say we merge this after a rebase, and open a new issue and PR to discuss adaptation issues. Do you agree?

**Comment by cpfiffer** [comment]
Agreed, I'll open a new issue.

**Pull request merged**